src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -266,18 +268,15 @@ protected virtual void OnConsumedBytes(int count)
                 return new ForChunkedEncoding(keepAlive, headers, context);
             }
 
-            var unparsedContentLength = headers.HeaderContentLength;
-            if (unparsedContentLength.Count > 0)
+            if (headers.ContentLength.HasValue)
```

Comment:
Is now (was throwing wrong exception, had to split request/response); is thrown during header parsing rather than message body - have added test.